File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */

const dfs = (start, candidates, target, res, path) => {
    if (target < 0) return;
    if (target === 0) {
        res.push([...path]);
        return;
    }
    for (let i = start; i < candidates.length; i++) {
        const num = candidates[i];
        path.push(num);
        dfs(i, candidates, target - num, res, path);
        path.pop();
    }
}

const combinationSum = (candidates, target) => {
    const res = [];
    dfs(0, candidates, target, res, []);
    return res;
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtrack(index, path, res, running_sum):
            if running_sum == target:
                res.append(path.copy())
                return
            if running_sum > target:
                return
            for i in range(index, len(candidates)):
                num = candidates[i]
                running_sum += num
                path.append(num)
                backtrack(index, path, res, running_sum)
                running_sum -= num
                path.pop()
                index += 1
        res = []
        backtrack(0, [], res, 0)
        return res